scipy/optimize/_linprog_util.py

```diff
@@ -128,16 +180,20 @@ def _clean_inputs(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None, bounds=None,
         Starting values of the independent variables, which will be refined by
         the optimization algorithm.
     """
+    if c is None:
+        raise TypeError
 
     try:
-        if c is None:
-            raise TypeError
-        try:
-            c = np.asarray(c, dtype=float).copy().squeeze()
-        except BaseException:  # typically a ValueError and shouldn't be, IMO
-            raise TypeError
+        c = np.array(c, dtype=np.float, copy=True).squeeze()
+    except ValueError:
+        raise TypeError(
+            "Invalid input for linprog: c must be a 1D array of numerical "
+            "coefficients")
+    else:
```

Comment:
Functionally these are equivalent, unless you were to also use the [finally](https://docs.python.org/3/reference/compound_stmts.html#finally) clause.

I added ``else`` to group the tests in a more logical way without having to wrap all the checks in a function. I don't think that's what ``else`` is meant to be used for but it makes the code neater so I'll use it.

Reply:
That's actually precisely what `else` is for. Quoting R. Hettinger, 

> "It [else block] runs when there is no exception but before the finally-clause. Without the else-clause, the only option to run additional code before finalization would be the clumsy practice of adding the code to the try-clause. That is clumsy because it risks raising exceptions in code that wasn't intended to be protected by the try-block."

Though I think you can also wrap `c` with [`atleast_1d`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.atleast_1d.html) to skip the 1d check

Reply:
Thanks @ilayn. I'll add that in.